src/metabase/query_processor/pivot.clj

```diff
@@ -1,98 +1,144 @@
-                              (vec (range 0 (count breakouts)))))))))
+    (group-number 3 [1])   ; -> 5
+    (group-number 3 [1 2]) ; -> 1"
+  [num-breakouts indecies]
+  (transduce
+   (map (partial bit-shift-left 1))
+   (completing bit-xor)
+   (int (dec (Math/pow 2 num-breakouts)))
+   indecies))
 
-(defn add-grouping-field
+(defn- breakout-combinations
+  "Return a sequence of all breakout combinations (by index) we should generate queries for.
+
+    (breakout-combinations 3 [1 2] nil) ;; -> [[0 1 2] [] [1 2] [2] [1]]"
+  [num-breakouts pivot-rows pivot-cols]
+  ;; validate pivot-rows/pivot-cols
+  (doseq [[k pivots] {:pivot-rows pivot-rows, :pivot-cols pivot-cols}
+          i          pivots]
+    (when (>= i num-breakouts)
+      (throw (ex-info (tru "Invalid {0}: specified breakout at index {1}, but we only have {2} breakouts"
+                           (name k) i num-breakouts)
+                      {:type          qp.error-type/invalid-query
+                       :num-breakouts num-breakouts
+                       :pivot-rows    pivot-rows
+                       :pivot-cols    pivot-cols}))))
+  (sort-by
+   (partial group-number num-breakouts)
+   (distinct
+    (map
+     vec
+     (concat
+      ;; e.g. given num-breakouts = 4; pivot-rows = [0 1 2]; pivot-cols = [3]
+      ;; subtotal rows
+      ;; _.range(1, pivotRows.length).map(i => [...pivotRow.slice(0, i), ...pivotCols])
+      ;;  => [0 _ _ 3] [0 1 _ 3] => 0110 0100 => Group #6, #4
+      (for [i (range 1 (count pivot-rows))]
+        (concat (take i pivot-rows) pivot-cols))
+      ;; “row totals” on the right
+      ;; pivotRows
+      ;; => [0 1 2 _] => 1000 => Group #8
+      [pivot-rows]
+      ;; subtotal rows within “row totals”
+      ;; _.range(1, pivotRows.length).map(i => pivotRow.slice(0, i))
+      ;; => [0 _ _ _] [0 1 _ _] => 1110 1100 => Group #14, #12
+      (for [i (range 1 (count pivot-rows))]
+        (take i pivot-rows))
+      ;; “grand totals” row
+      ;; pivotCols
+      ;; => [_ _ _ 3] => 0111 => Group #7
+      [pivot-cols]
+      ;; bottom right corner [_ _ _ _] => 1111 => Group #15
+      [[]])))))
+
+(s/defn ^:private add-grouping-field
   "Add the grouping field and expression to the query"
-  [query breakout bitmask]
-  (let [new-query (-> query
+  [outer-query breakouts group-number :- s/Int]
+  (let [new-query (-> outer-query
                       ;;TODO: `pivot-grouping` is not "magic" enough to mark it as an internal thing
                       (update-in [:query :fields]
                                  #(conj % [:expression "pivot-grouping"]))
-                      ;;TODO: replace this value with a bitmask or something to indicate the source better
                       (update-in [:query :expressions]
-                                 #(assoc % "pivot-grouping" [:abs bitmask])))]
+                                 #(assoc % "pivot-grouping" [:abs group-number])))]
```

Comment:
This is a bitmask, per https://www.postgresql.org/docs/current/queries-table-expressions.html#QUERIES-GROUPING-SETS at Paul's request, hence the name `bitmask`

Reply:
gotcha... I renamed it back to `group-bitmask`